spec/acceptance/ios_snmp_trap.rb

```diff
@@ -0,0 +1,87 @@
+require 'spec_helper_acceptance'
+
+describe 'ios_snmp_trap' do
+  target = if ['3650'].include?(device_model)
+             'GigabitEthernet1/0'
+           else
+             'GigabitEthernet0'
+           end
+  it 'Set Two Instances' do
+    pp = <<-EOS
+    ios_snmp_trap { '#{target}/1':
+      link_status => true,
+    }
+    ios_snmp_trap { '#{target}/2':
+      mac_notification_added => true,
+      mac_notification_removed => true,
+    }
+    EOS
+    make_site_pp(pp)
+    run_device(allow_changes: true)
+    # Are we idempotent
+    run_device(allow_changes: false)
+    # Check puppet resource
+    result = run_resource('ios_snmp_trap', "#{target}/1")
+    expect(result).to match(%r{mac_notification_added => false})
+    expect(result).to match(%r{mac_notification_removed => false})
+    expect(result).to match(%r{link_status => true})
+    result = run_resource('ios_snmp_trap', "#{target}/2")
+    expect(result).to match(%r{mac_notification_added => true})
+    expect(result).to match(%r{mac_notification_removed => true})
+    expect(result).to match(%r{link_status => false})
+  end
+
+  it 'Update Two Instances' do
+    pp = <<-EOS
+    ios_snmp_trap { '#{target}/1':
+      mac_notification_added => true,
+      mac_notification_removed => true,
+      link_status => false,
+    }
+    ios_snmp_trap { '#{target}/2':
+      mac_notification_added => false,
+      mac_notification_removed => false,
+      link_status => true,
+    }
+    EOS
+    make_site_pp(pp)
+    run_device(allow_changes: true)
+    # Are we idempotent
+    run_device(allow_changes: false)
+    # Check puppet resource
+    result = run_resource('ios_snmp_trap', "#{target}/1")
+    expect(result).to match(%r{mac_notification_added => true})
+    expect(result).to match(%r{mac_notification_removed => true})
+    expect(result).to match(%r{link_status => false})
+    result = run_resource('ios_snmp_trap', "#{target}/2")
+    expect(result).to match(%r{mac_notification_added => false})
+    expect(result).to match(%r{mac_notification_removed => false})
+    expect(result).to match(%r{link_status => true})
+  end
+
+  it 'Unset Two Instances' do
+    pp = <<-EOS
+    ios_snmp_trap { '#{target}/1':
+      mac_notification_added => false,
+      mac_notification_removed => false,
+      link_status => false,
+    }
+    ios_snmp_trap { '#{target}/2':
+      mac_notification_added => false,
+      mac_notification_removed => false,
+      link_status => false,
+    }
+    EOS
+    make_site_pp(pp)
+    run_device(allow_changes: true)
+    # Check puppet resource
```

Comment:
Guessing this is not considered idempotent as there is no idempotency check? Could it potentially be updated to be idempotent, i.e. if the `get value` is not present then I think it is safe to assume that `mac_notification_added` is false. This might be a case of just not having them as `Optionals`, and defaults if it makes sense - the `enforce_simple_types` should handle the rest

Reply:
Strangely, having an idempotent check here causes the update test to fail an idempotent check.  No idea why

Reply:
Not present does equal false, it's just that the setting false is converted to unset

Reply:
Ah, finally remembered. This is due to how the code works, Whenever the resources are retrieved it grabs every single port so they are all displayed. I altered the code so that it only shows ones that have at least one attribute set to true in order to keep it neater.

Reply:
So when it goes looking for the port for the second apply it can't find it.

Reply:
So this goes back to the comment from @shermdog about having it as a larger `ios_interface` type/provider for managing layer 3 configuration such as this and not be snmp_trap specific, as there are other layer 3 commands that will need to be added in future - since this is showing features enabled of an interface, it should still show if the feature is turned on or not in my opinion. I would visualise that it would be something like:

```
ios_interface { 'GigabitEthernet1/1':
  mac_notification_added => false,
  mac_notification_removed => false,
  link_status => false
}
...
ios_interface { 'GigabitEthernet1/64':
  mac_notification_added => true,
  mac_notification_removed => false,
  link_status => true
}
```

This way the resource is considered idempotent, because at the minute it will not be with something like:

```
ios_interface { 'GigabitEthernet1/1':
  mac_notification_added => false,
  mac_notification_removed => false,
  link_status => false
}
```

Reply:
Should this be merged in with `network_interface` then???

Reply:
Since it is already sort of this?

Reply:
No, `network_interface` is more just a `enable/disable this interface`, while `ios_interface` would be more of the layer 3 configuration settings which we currently do not model as `snmp trap` is our first use case based on the limited customer configs we have.

Reply:
Ok, have converted it to be called ios_interface and removed the code that cut the results down to only the ports that had something in them set.